### PR TITLE
refactor(api): delete lastUpdate and add missing events

### PR DESCRIPTION
**Pull request by bnjjj**
Signed-off-by: Benjamin Coenen <benjamin.coenen@corp.ovh.com>

1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/api/application/application_group.go [hunk 1]
```diff
@@ -112,10 +111,7 @@ func AddGroup(db gorp.SqlExecutor, store cache.Store, proj *sdk.Project, a *sdk.
 			if err := group.InsertGroupInProject(db, proj.ID, g.ID, perm); err != nil {
 				return sdk.WrapError(err, "AddGroup> Cannot add group %s in project %s", g.Name, proj.Name)
 			}
-
-			if err := UpdateLastModified(db, store, a, u); err != nil {
-				return sdk.WrapError(err, "AddGroup> Cannot update application %s", a.Name)
-			}
+			event.PublishAddProjectPermission(proj, sdk.GroupPermission{Group: *g, Permission: perm}, u)
```
Comment (sguiheux) [inline engine/api/application/application_group.go]: remove

File: engine/api/application/application_group.go [hunk 2]
```diff
@@ -131,9 +127,7 @@ func AddGroup(db gorp.SqlExecutor, store cache.Store, proj *sdk.Project, a *sdk.
 					return sdk.WrapError(err, "AddGroup> Cannot add group %s in pipeline %s", g.Name, p.Pipeline.Name)
 				}
 
-				if err := pipeline.UpdatePipelineLastModified(db, store, proj, &p.Pipeline, u); err != nil {
-					return sdk.WrapError(err, "AddGroup> Cannot update pipeline %s", p.Pipeline.Name)
-				}
+				event.PublishPipelinePermissionAdd(proj.Key, p.Pipeline.Name, sdk.GroupPermission{Group: *g, Permission: perm}, u)
```
Comment (sguiheux) [inline engine/api/application/application_group.go]: remove

File: engine/api/application/application_group.go [hunk 3]
```diff
@@ -149,9 +143,7 @@ func AddGroup(db gorp.SqlExecutor, store cache.Store, proj *sdk.Project, a *sdk.
 							return sdk.WrapError(err, "AddGroup> Cannot add group %s in env %s", g.Name, t.DestEnvironment.Name)
 						}
 
-						if err := environment.UpdateLastModified(db, store, u, &t.DestEnvironment); err != nil {
-							return sdk.WrapError(err, "AddGroup> Cannot update env %s", t.DestEnvironment.Name)
-						}
+						event.PublishEnvironmentPermissionAdd(proj.Key, t.DestEnvironment, sdk.GroupPermission{Group: *g, Permission: perm}, u)
```
Comment (sguiheux) [inline engine/api/application/application_group.go]: remove

File: engine/api/application/application_group.go [hunk 4]
```diff
@@ -165,9 +157,7 @@ func AddGroup(db gorp.SqlExecutor, store cache.Store, proj *sdk.Project, a *sdk.
 							return sdk.WrapError(err, "AddGroup> Cannot add group %s in env %s", g.Name, t.SrcEnvironment.Name)
 						}
 
-						if err := environment.UpdateLastModified(db, store, u, &t.SrcEnvironment); err != nil {
-							return sdk.WrapError(err, "AddGroup> Cannot update env %s", t.SrcEnvironment.Name)
-						}
+						event.PublishEnvironmentPermissionAdd(proj.Key, t.SrcEnvironment, sdk.GroupPermission{Group: *g, Permission: perm}, u)
```
Comment (sguiheux) [inline engine/api/application/application_group.go]: remove

File: engine/api/application/application_parser.go [hunk 5]
```diff
@@ -177,6 +178,7 @@ func ParseAndImport(db gorp.SqlExecutor, cache cache.Store, proj *sdk.Project, e
 	globalError := Import(db, cache, proj, app, eapp.VCSServer, u, msgChan)
 	close(msgChan)
 	done.Wait()
+	event.PublishAddApplication(proj.Key, *app, u)
```
Comment (sguiheux) [inline engine/api/application/application_parser.go]: remove

File: engine/api/application/application_variable.go [hunk 6]
```diff
@@ -240,8 +241,9 @@ func InsertVariable(db gorp.SqlExecutor, store cache.Store, app *sdk.Application
 	if err := inserAudit(db, ava); err != nil {
 		return sdk.WrapError(err, "InsertVariable> Cannot insert audit for variable %d", variable.ID)
 	}
+	event.PublishAddVariableApplication(app.ProjectKey, *app, variable, u)
```
Comment (sguiheux) [inline engine/api/application/application_variable.go]: remove

File: engine/api/application/application_variable.go [hunk 7]
```diff
@@ -286,9 +288,9 @@ func UpdateVariable(db gorp.SqlExecutor, store cache.Store, app *sdk.Application
 	if err := inserAudit(db, ava); err != nil {
 		return sdk.WrapError(err, "UpdateVariable> Cannot insert audit for variable %s", variable.Name)
 	}
+	event.PublishUpdateVariableApplication(app.ProjectKey, *app, *variable, *variableBefore, u)
```
Comment (sguiheux) [inline engine/api/application/application_variable.go]: remove

File: engine/api/application/application_variable.go [hunk 8]
```diff
@@ -320,8 +322,9 @@ func DeleteVariable(db gorp.SqlExecutor, store cache.Store, app *sdk.Application
 	if err := inserAudit(db, ava); err != nil {
 		return sdk.WrapError(err, "DeleteVariable> Cannot insert audit for variable %s", variable.Name)
 	}
+	event.PublishDeleteVariableApplication(app.ProjectKey, *app, *variable, u)
```
Comment (sguiheux) [inline engine/api/application/application_variable.go]: remove

File: engine/api/application/dao.go [hunk 9]
```diff
@@ -294,7 +294,9 @@ func Insert(db gorp.SqlExecutor, store cache.Store, proj *sdk.Project, app *sdk.
 		return sdk.WrapError(err, "application.Insert %s(%d)", app.Name, app.ID)
 	}
 	*app = sdk.Application(dbApp)
-	return UpdateLastModified(db, store, app, u)
+	event.PublishAddApplication(proj.Key, *app, u)
```
Comment (sguiheux) [inline engine/api/application/dao.go]: remove

File: engine/api/application_key.go [hunk 10]
```diff
@@ -60,9 +61,7 @@ func (api *API) deleteKeyInApplicationHandler() Handler {
 				if err := application.DeleteApplicationKey(tx, app.ID, keyName); err != nil {
 					return sdk.WrapError(err, "deleteKeyInApplicationHandler> Cannot delete key %s", k.Name)
 				}
-				if err := application.UpdateLastModified(tx, api.Cache, app, getUser(ctx)); err != nil {
-					return sdk.WrapError(err, "deleteKeyInApplicationHandler> Cannot update application last modified date")
-				}
+				event.PublishApplicationKeyDelete(key, *app, keyToDelete, u)
```
Comment (sguiheux) [inline engine/api/application_key.go]: after commit

File: engine/api/application_migration.go [hunk 11]
```diff
@@ -41,26 +44,23 @@ func (api *API) migrationApplicationWorkflowCleanHandler() Handler {
...
 		p.WorkflowMigration = migrate.STATUS_START
-		if err := project.Update(tx, api.Cache, p, getUser(ctx)); err != nil {
-			return sdk.WrapError(err, "migrationApplicationWorkflowHandler")
-		}
-
-		if err := project.UpdateLastModified(tx, api.Cache, getUser(ctx), p, sdk.ProjectLastModificationType); err != nil {
+		if err := project.Update(tx, api.Cache, p, u); err != nil {
 			return sdk.WrapError(err, "migrationApplicationWorkflowHandler")
 		}
+		event.PublishUpdateProject(p, &oldProj, u)
```
Comment (sguiheux) [inline engine/api/application_migration.go]: after commit

File: engine/api/environment/environment_importer.go [hunk 12]
```diff
@@ -70,6 +71,8 @@ func Import(db gorp.SqlExecutor, proj *sdk.Project, env *sdk.Environment, msgCha
 		msgChan <- sdk.NewMessage(sdk.MsgEnvironmentCreated, env.Name)
 	}
 
+	event.PublishEnvironmentAdd(proj.Key, *env, u)
```
Comment (sguiheux) [inline engine/api/environment/environment_importer.go]: remove

File: engine/api/environment/environment_importer.go [hunk 13]
```diff
@@ -191,5 +194,7 @@ func ImportInto(db gorp.SqlExecutor, proj *sdk.Project, env *sdk.Environment, in
 
 	log.Debug("ImportInto> Done")
 
+	event.PublishEnvironmentUpdate(proj.Key, *env, *into, u)
```
Comment (sguiheux) [inline engine/api/environment/environment_importer.go]: remove

File: engine/api/environment_group.go [hunk 14]
```diff
@@ -181,6 +155,8 @@ func (api *API) addGroupsInEnvironmentHandler() Handler {
 		envUpdated.Permission = permission.EnvironmentPermission(key, envUpdated.Name, getUser(ctx))
 		envUpdated.ProjectKey = key
 
+		event.PublishEnvironmentUpdate(key, *envUpdated, *env, getUser(ctx))
```
Comment (sguiheux) [inline engine/api/environment_group.go]: wrong event EventEnvironmentPermissionAdd

File: engine/api/group.go [hunk 15]
```diff
@@ -90,31 +92,21 @@ func (api *API) deleteGroupHandler() Handler {
 			return sdk.WrapError(err, "deleteGroupHandler> cannot delete group")
 		}
 
+		groupPerm := sdk.GroupPermission{Group: *g}
 		for _, pg := range projPerms {
-			if err := project.UpdateLastModified(tx, api.Cache, getUser(ctx), &pg.Project, sdk.ProjectLastModificationType); err != nil {
-				return sdk.WrapError(err, "deleteGroupHandler> Cannot update project last modified date")
-			}
+			event.PublishDeleteProjectPermission(&pg.Project, groupPerm, u)
```
Comment (sguiheux) [inline engine/api/group.go]: after commit

File: engine/api/group.go [hunk 16]
```diff
@@ -90,31 +92,21 @@ func (api *API) deleteGroupHandler() Handler {
...
 		}
 
 		for _, pg := range appPerms {
-			if err := application.UpdateLastModified(tx, api.Cache, &pg.Application, getUser(ctx)); err != nil {
-				return sdk.WrapError(err, "deleteGroupHandler> Cannot update application last modified date")
-			}
+			event.PublishApplicationPermissionDelete(pg.Application.ProjectKey, pg.Application, groupPerm, u)
```
Comment (sguiheux) [inline engine/api/group.go]: after commit

File: engine/api/group.go [hunk 17]
```diff
@@ -90,31 +92,21 @@ func (api *API) deleteGroupHandler() Handler {
...
 		}
 
 		for _, pg := range pipPerms {
-			p := &sdk.Project{
-				Key: pg.Pipeline.ProjectKey,
-			}
-			if err := pipeline.UpdatePipelineLastModified(tx, api.Cache, p, &pg.Pipeline, getUser(ctx)); err != nil {
-				return sdk.WrapError(err, "deleteGroupHandler> Cannot update pipeline last modified date")
-			}
+			event.PublishPipelinePermissionDelete(pg.Pipeline.ProjectKey, pg.Pipeline.Name, groupPerm, u)
```
Comment (sguiheux) [inline engine/api/group.go]: after commit

File: engine/api/group.go [hunk 18]
```diff
@@ -90,31 +92,21 @@ func (api *API) deleteGroupHandler() Handler {
...
 		}
 
 		for _, pg := range envPerms {
-			if err := environment.UpdateLastModified(tx, api.Cache, getUser(ctx), &pg.Environment); err != nil {
-				return sdk.WrapError(err, "deleteGroupHandler> Cannot update environment last modified date")
-			}
+			event.PublishEnvironmentPermissionDelete(pg.Environment.ProjectKey, pg.Environment, groupPerm, u)
```
Comment (sguiheux) [inline engine/api/group.go]: after commit

File: engine/api/project/dao.go [hunk 19]
```diff
@@ -194,8 +194,9 @@ func Insert(db gorp.SqlExecutor, store cache.Store, proj *sdk.Project, u *sdk.Us
 	if err := InsertKey(db, &pk); err != nil {
 		return sdk.WrapError(err, "project.Insert> Unable to insert PGPKeyPair")
 	}
+	event.PublishAddProject(proj, u)
```
Comment (sguiheux) [inline engine/api/project/dao.go]: remove

File: engine/api/status.go [hunk 20]
```diff
@@ -47,7 +47,7 @@ func (api *API) Status() sdk.MonitoringStatus {
 	m.Lines = append(m.Lines, getStatusLine(objectstore.Status()))
 	m.Lines = append(m.Lines, getStatusLine(mail.Status()))
 	m.Lines = append(m.Lines, getStatusLine(api.DBConnectionFactory.Status()))
-	m.Lines = append(m.Lines, getStatusLine(sdk.MonitoringStatusLine{Component: "LastUpdate Connected", Value: fmt.Sprintf("%d", len(api.lastUpdateBroker.clients)), Status: sdk.MonitoringStatusOK}))
+	m.Lines = append(m.Lines, getStatusLine(sdk.MonitoringStatusLine{Component: "SSE Connected", Value: fmt.Sprintf("%d", len(api.eventsBroker.clients)), Status: sdk.MonitoringStatusOK}))
```
Comment (sguiheux) [inline engine/api/status.go]: maybe a pb of concurrent access on  this map ?

File: engine/api/group.go (hunk 17 above)
Comment (bnjjj) [reply engine/api/group.go]: done

File: engine/api/group.go (hunk 16 above)
Comment (bnjjj) [reply engine/api/group.go]: done

File: engine/api/group.go (hunk 18 above)
Comment (bnjjj) [reply engine/api/group.go]: done

**Comment by ovh-cds** [comment]
CDS Report ut-engine#5293.0 ✘  
* Stage 1
  * Engine Test with DB ✘  
Unit Tests Report
* github.com/ovh/cds/cli


* github.com/ovh/cds/engine/api






































































































































































































  * Test_postWorkflowHandlerWithRootShouldSuccess ✘

* github.com/ovh/cds/engine/api/application








* github.com/ovh/cds/engine/api/environment




* github.com/ovh/cds/engine/api/keys



* github.com/ovh/cds/engine/api/migrate


* github.com/ovh/cds/engine/api/pipeline






* github.com/ovh/cds/engine/api/plugin


* github.com/ovh/cds/engine/api/poller




* github.com/ovh/cds/engine/api/project









* github.com/ovh/cds/engine/api/scheduler












* github.com/ovh/cds/engine/api/secret






* github.com/ovh/cds/engine/api/user








* github.com/ovh/cds/engine/api/warning















* github.com/ovh/cds/engine/api/worker










* github.com/ovh/cds/engine/api/workflow













































* github.com/ovh/cds/engine/hatchery/swarm



















* github.com/ovh/cds/engine/hooks






* github.com/ovh/cds/engine/repositories


* github.com/ovh/cds/engine/vcs











* github.com/ovh/cds/engine/vcs/bitbucket


















* github.com/ovh/cds/engine/vcs/github













* github.com/ovh/cds/engine/vcs/gitlab









* github.com/ovh/cds/engine/worker






* github.com/ovh/cds/sdk















* github.com/ovh/cds/sdk/exportentities




























* github.com/ovh/cds/sdk/interpolate

























* github.com/ovh/cds/sdk/luascript





* github.com/ovh/cds/sdk/plugin/dummy/caller


* github.com/ovh/cds/sdk/vcs/git

**Pull request merged**